### PR TITLE
[FLINK-17678][hbase] Support fink-sql-connector-hbase

### PR DESCRIPTION
**Pull request by leonardBang**
## What is the purpose of the change

* Flink doesn't contains a hbase shade jar right now, so users have to add hbase dependency manually. This pull request import a new flink-sql-connector-hbase module likes flink-sql-connector-elasticsearch7 to offer bundled jar.

* **Specially, this PR is inspired by #12369， #9898** 

## Brief change log

- Add new modular flink-sql-connector-hbase.
- Add new modular flink-end-to-end-test-hbase.

## Verifying this change

* End2end test `SQLClientHbaseITCase` covered
The new module flink-sql-connector-hbase only includes necessary dependencies that used to reading or writing data from/to hbase. And the module flink-end-to-end-test-hbase, just as name implies, is used for hbase e2e test.

(1) We do not shade hadoop dependencies into the jar. Because flink-sql-connector-hive doesn't contains either, so we do this in similar way.

(2) We shade all dependencies that hbase needed but except org.apache.hadoop.hbase.codec.* for preventing hbase region server throw timeout exception . Because the hbase region server can not find the shaded codec class to decoding data(byte[]).

(3) We add a new module, flink-end-to-end-test-hbase, only for e2e test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+- commons-lang:commons-lang:2.6
+- commons-logging:commons-logging:1.1.3
+- com.google.guava:guava:12.0.1
+- com.yammer.metrics:metrics-core:2.2.0
+- io.netty:netty-all:4.1.44.Final
+
+This project bundles the following dependencies under the BSD license.
+See bundled license files for details.
+
+- com.google.protobuf:protobuf-java:2.5.0
```
Comment (JingsongLi) [inline flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:27]: Add license file for protobuf

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE [hunk 2]
```diff
@@ -0,0 +1,27 @@
+flink-sql-connector-hbase
+Copyright 2014-2020 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
+
+This project bundles the following dependencies under the Apache Software License 2.0. (http://www.apache.org/licenses/LICENSE-2.0.txt)
+
+- org.apache.hbase:hbase-common:1.4.3
```
Comment (JingsongLi) [inline flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:9]: sort them by order

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+- commons-logging:commons-logging:1.1.3
+- com.google.guava:guava:12.0.1
+- com.yammer.metrics:metrics-core:2.2.0
+- io.netty:netty-all:4.1.44.Final
+- org.apache.hbase:hbase-common:1.4.3
+- org.apache.hbase:hbase-protocol:1.4.3
+- org.apache.hbase:hbase-procedure:1.4.3
+- org.apache.hbase:hbase-client:1.4.3
+- org.apache.hbase:hbase-prefix-tree:1.4.3
+- org.apache.htrace:htrace-core:3.1.0-incubating
```
Comment (wuchong) [inline flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:21]: htrace-core bundles several dependencies but doesn't have a proper NOTICE. Maybe we should list the dependencies:

```
htrace-core
From: 'FasterXML' (http://fasterxml.com/)
  - Jackson-annotations (http://wiki.fasterxml.com/JacksonHome) com.fasterxml.jackson.core:jackson-annotations:bundle:2.4.0
    License: The Apache Software License, Version 2.0  (http://www.apache.org/licenses/LICENSE-2.0.txt)
  - Jackson-core (http://wiki.fasterxml.com/JacksonHome) com.fasterxml.jackson.core:jackson-core:bundle:2.4.0
    License: The Apache Software License, Version 2.0  (http://www.apache.org/licenses/LICENSE-2.0.txt)
  - jackson-databind (http://wiki.fasterxml.com/JacksonHome) com.fasterxml.jackson.core:jackson-databind:bundle:2.4.0
    License: The Apache Software License, Version 2.0  (http://www.apache.org/licenses/LICENSE-2.0.txt)

From: 'The Apache Software Foundation' (http://www.apache.org/)
  - Commons Logging (http://commons.apache.org/logging) commons-logging:commons-logging:jar:1.1.1
    License: The Apache Software License, Version 2.0  (http://www.apache.org/licenses/LICENSE-2.0.txt)
```

The `commons-logging` is a different version.

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+This project bundles the following dependencies under the Apache Software License 2.0. (http://www.apache.org/licenses/LICENSE-2.0.txt)
+
+- commons-codec:commons-codec:1.10
+- commons-configuration:commons-configuration:1.7
+- commons-lang:commons-lang:2.6
+- commons-logging:commons-logging:1.1.3
+- com.google.guava:guava:12.0.1
+- com.yammer.metrics:metrics-core:2.2.0
+- io.netty:netty-all:4.1.44.Final
```
Comment (wuchong) [inline flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:15]: Shall we list all the netty dependencies if `netty-all` is included? 

```
io.netty:netty-codec-smtp:4.1.44.Final
io.netty:netty-transport-native-epoll:4.1.44.Final
io.netty:netty-codec-redis:4.1.44.Final
io.netty:netty-resolver-dns:4.1.44.Final
io.netty:netty-codec-dns:4.1.44.Final
io.netty:netty-resolver-dns-native-macos:4.1.44.Final
io.netty:netty-codec-mqtt:4.1.44.Final
io.netty:netty-codec-http2:4.1.44.Final
io.netty:netty-transport-native-unix-common:4.1.44.Final
io.netty:netty-transport-sctp:4.1.44.Final
io.netty:netty-transport-native-epoll:4.1.44.Final
io.netty:netty-codec-http:4.1.44.Final
io.netty:netty-transport-udt:4.1.44.Final
io.netty:netty-handler-proxy:4.1.44.Final
io.netty:netty-resolver:4.1.44.Final
io.netty:netty-common:4.1.44.Final
io.netty:netty-codec-xml:4.1.44.Final
io.netty:netty-transport:4.1.44.Final
io.netty:netty-codec:4.1.44.Final
io.netty:netty-codec-socks:4.1.44.Final
io.netty:netty-codec-stomp:4.1.44.Final
io.netty:netty-buffer:4.1.44.Final
io.netty:netty-transport-rxtx:4.1.44.Final
io.netty:netty-codec-haproxy:4.1.44.Final
io.netty:netty-codec-haproxy:4.1.44.Final
io.netty:netty-tcnative:2.0.28.Final
io.netty:netty-transport-native-kqueue:4.1.44.Final
io.netty:netty-handler:4.1.44.Final
io.netty:netty-codec-memcache:4.1.44.Final
```

I'm not sure about this, but as there are no other 3rd party dependencies in `netty-all`, maybe it's fine just include `netty-all`.

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE (hunk 3 above)
Comment (leonardBang) [reply flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:21]: I agree we should list all bundled dependencies by htrace-core, and also need to add different version of `commons-logging`

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE (hunk 4 above)
Comment (leonardBang) [reply flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:15]: cc @JingsongLi Could you help confirm this?

Comment (JingsongLi) [reply flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:15]: Maybe we need, see hadoop notice https://cwiki.apache.org/confluence/display/HADOOP/Bundled+dependencies

**Comment by leonardBang** [comment]
@rmetzger Could you help take a look the NOTICE of the `flink-sql-connector-hbase` module ? many appreciate.

**Comment by rmetzger** [comment]
Yes, I will take a look later today

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE (hunk 4 above)
Comment (rmetzger) [reply flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:19]: I believe we don't need to "expand" the dependencies included in "netty-all". Everything that is included in netty-all is licensed to the netty project (as long as there's nothing shaded etc.)

File: flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
+
+This project bundles the following dependencies under the Apache Software License 2.0. (http://www.apache.org/licenses/LICENSE-2.0.txt)
+
+- commons-codec:commons-codec:1.10
+- commons-configuration:commons-configuration:1.7
+- commons-lang:commons-lang:2.6
+- commons-logging:commons-logging:1.1.1
```
[the remark this answers is not in the document]
Comment (leonardBang) [reply flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE:12]: maven shade plugin output may miss them?  These dependencies
```
com.fasterxml.jackson.core:jackson-annotations:2.4.0
com.fasterxml.jackson.core:jackson-core:2.4.0 
com.fasterxml.jackson.core:jackson-databind:2.4.0
commons-logging:commons-logging:1.1.1
```
has been shaded in `htrace-core`, I found the dependencies version in maven shade plugin output are different with these dependencies in `META-INF/DEPENDENCIES` of `htrace-core-3.1.0-incubating.jar`

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-connectors/flink-sql-connector-hbase/src/main/resources/META-INF/NOTICE]: `htrace-core` is an uber jar which shades `jackson` and `commons-logging` in it. I think that's why we can't find the dependencies in maven shade plugin output?

File: flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/LocalStandaloneFlinkResourceFactory.java [hunk 6]
```diff
@@ -80,6 +80,33 @@ public FlinkResource create(FlinkResourceSetup setup) {
...
+		} else {
+			// running in the IDE; working directory is test module
+			Optional<Path> projectRootDirectory = findProjectRootDirectory(currentDirectory);
+			// this distinction is required in case this class is used outside of Flink
+			if (projectRootDirectory.isPresent()) {
+				projectRootPath = projectRootDirectory.get();
+			} else {
+				throw new IllegalArgumentException(
+					"The 'rootDir' property was not set and the flink project root directory could not be found" +
+						" automatically. Please point the 'rootDir' property to the  flink project root directory;" +
```
Comment (wuchong) [inline flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/LocalStandaloneFlinkResourceFactory.java:103]:
```suggestion
						" automatically. Please point the 'rootDir' property to the flink project root directory;" +
```

File: flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/LocalStandaloneFlinkResourceFactory.java [hunk 7]
```diff
@@ -80,6 +80,33 @@ public FlinkResource create(FlinkResourceSetup setup) {
 		return new LocalStandaloneFlinkResource(distributionDirectory.get(), logBackupDirectory.orElse(null), setup);
 	}
 
+	/**
+	 * Utils to find the flink project root directory.
+	 * @param currentDirectory
+	 * @return The flink project root directory.
+	 */
+	public static Path getProjectRootDirectory(Path currentDirectory) {
```
Comment (wuchong) [inline flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/LocalStandaloneFlinkResourceFactory.java:88]: We can use this method to replace the code block in `LocalStandaloneFlinkResourceFactory#create()`.

**Comment by leonardBang** [comment]
azure pipeline passed, failed test is not related to this PR
 https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=3783&view=logs&j=c88eea3b-64a0-564d-0031-9fdcd7b8abee&t=1e2bbe5b-4657-50be-1f07-d84bfce5b1f5

**Comment by leonardBang** [comment]
> Could you rebase and squash to trigger the build again?

Sure, squashed and triggered

**Comment by wuchong** [comment]
Build is passed, merging...

**Pull request merged**